auditlog/formats.go

```diff
@@ -21,59 +21,79 @@ package auditlog
 
 import (
 	"fmt"
+	"strings"
 
 	utils "github.com/corazawaf/coraza/v3/internal/strings"
+	"github.com/corazawaf/coraza/v3/types"
 )
 
 func nativeFormatter(al *Log) ([]byte, error) {
-	boundary := utils.RandomString(10)
-	parts := map[byte]string{}
-	// [27/Jul/2016:05:46:16 +0200] V5guiH8AAQEAADTeJ2wAAAAK 192.168.3.1 50084 192.168.3.111 80
-	parts['A'] = fmt.Sprintf("[%s] %s %s %d %s %d", al.Transaction.Timestamp, al.Transaction.ID,
-		al.Transaction.ClientIP, al.Transaction.ClientPort, al.Transaction.HostIP, al.Transaction.HostPort)
-	// GET /url HTTP/1.1
-	// Host: example.com
-	// User-Agent: Mozilla/5.0
-	// Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
-	// Accept-Language: en-US,en;q=0.5
-	// Accept-Encoding: gzip, deflate
-	// Referer: http://example.com/index.html
-	// Connection: keep-alive
-	// Content-Type: application/x-www-form-urlencoded
-	// Content-Length: 6
-	parts['B'] = fmt.Sprintf("%s %s %s\n", al.Transaction.Request.Method, al.Transaction.Request.URI, al.Transaction.Request.Protocol)
-	for k, vv := range al.Transaction.Request.Headers {
-		for _, v := range vv {
-			parts['B'] += fmt.Sprintf("%s: %s\n", k, v)
-		}
-	}
-	// b=test
-	parts['C'] = al.Transaction.Request.Body
-	parts['E'] = al.Transaction.Response.Body
-	parts['F'] = ""
-	for k, vv := range al.Transaction.Response.Headers {
-		for _, v := range vv {
-			parts['F'] += fmt.Sprintf("%s: %s\n", k, v)
+	boundaryPrefix := fmt.Sprintf("--%s-", utils.RandomString(10))
+
+	var res strings.Builder
+
+	for _, part := range al.Parts {
+		res.WriteString(boundaryPrefix)
+		res.WriteByte(byte(part))
+		res.WriteString("--\n")
+		switch part {
+		case types.AuditLogPartAuditLogHeader:
+			// [27/Jul/2016:05:46:16 +0200] V5guiH8AAQEAADTeJ2wAAAAK 192.168.3.1 50084 192.168.3.111 80
+			_, _ = fmt.Fprintf(&res, "[%s] %s %s %d %s %d", al.Transaction.Timestamp, al.Transaction.ID,
+				al.Transaction.ClientIP, al.Transaction.ClientPort, al.Transaction.HostIP, al.Transaction.HostPort)
+		case types.AuditLogPartRequestHeaders:
+			// GET /url HTTP/1.1
+			// Host: example.com
+			// User-Agent: Mozilla/5.0
+			// Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
+			// Accept-Language: en-US,en;q=0.5
+			// Accept-Encoding: gzip, deflate
+			// Referer: http://example.com/index.html
+			// Connection: keep-alive
+			// Content-Type: application/x-www-form-urlencoded
+			// Content-Length: 6
+			_, _ = fmt.Fprintf(&res, "%s %s %s\n", al.Transaction.Request.Method, al.Transaction.Request.URI, al.Transaction.Request.Protocol)
+			for k, vv := range al.Transaction.Request.Headers {
+				for _, v := range vv {
+					res.WriteString(k)
+					res.WriteString(": ")
+					res.WriteString(v)
+					res.WriteByte('\n')
+				}
+			}
+		case types.AuditLogPartRequestBody:
+			// b=test
+			res.WriteString(al.Transaction.Request.Body)
+		case types.AuditLogPartIntermediaryResponseBody:
```

Comment:
This is the same as the current code but it doesn't look right to me

Reply:
Are you wondering about `AuditLogPartIntermediaryResponseBody ` vs `AuditLogPartResponseBody`? I guess that the latter was meant to log the response body after potential changes to it from Coraza? https://github.com/corazawaf/coraza/blob/31ffef063e2ded57c5fa8e91bf9eebd8458f65f2/types/waf.go#L122